### PR TITLE
Phase one

### PR DESCRIPTION
**Pull request by AshwinB-hat**

### CODE REVIEW

File: eli5/_shap_explanation_summary.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
+from eli5._feature_weights import get_top_features_filtered
+from eli5.base import Explanation, TargetExplanation
+from eli5._feature_importances import get_feature_importance_explanation
+
+DECISION_PATHS_DESCRIPTION = """
+Feature weights are calculated by following decision paths in trees
```
Comment (ivanprado) [inline eli5/_shap_explanation_summary.py]: I think all this unused descriptions and code is because you copy&paste the code from elsewhere and then you worked with it. Could you please remove this code when commiting? It would simplify my review labour :-) In general try to have your code as clean as possible also during development, as it is a good practice.

File: eli5/shap_xgboost.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+            raise ValueError(
+                'Inconsistent is_regression={} passed. '
+                'You don\'t have to pass it when using scikit-learn API'
+                .format(is_regression))
+        is_regression = _is_regression
+    return booster, is_regression
+
+def _xgb_feature_importances(booster, dMatrix):
+    fs = booster.predict(dMatrix, pred_contribs=True)
+    all_features = np.sum(fs, axis=0)
```
Comment (ivanprado) [inline eli5/shap_xgboost.py]: Consider use `np.mean(fs, axis=0)` and remove the division.

File: eli5/shap_xgboost.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    
+    if str(type(doc)).endswith("xgboost.core.DMatrix'>"):
+        dmatrix = doc
+    else:
+        if(vec is not None):
+            dmatrix=DMatrix(get_X(doc,vec=vec,vectorized=vectorized), missing=missing)
+        else:
+            dmatrix = DMatrix(doc, missing=missing)
+    coef = _xgb_feature_importances(booster, dmatrix)
+    return get_shap_explanation_summary(xgb, vec=vec, coef=coef,
```
Comment (ivanprado) [inline eli5/shap_xgboost.py]: I think I said you the other day something wrong. I said that you shouldn't use `get_decision_path_explanation`, but I was wrong. Probably you should use it here instead of `get_shap_explanation_summary` as was used here: https://github.com/AshwinB-hat/eli5/blob/master/eli5/xgboost.py#L201. Some ideas (I could be wrong):
- `explain_shap_prediction` in eli5 only supports a single sample in doc. It is a limitation of Eli5. It would be nice to support explanations of multiple documents at the same time, but forget about that by now.
- `score_weights` (https://github.com/AshwinB-hat/eli5/blob/master/eli5/xgboost.py#L195) in your case would be a mix between the output of `_xgb_feature_importances` and the predictions for the document, but the prediction for the document is the sum of the feature importances according to https://github.com/dmlc/xgboost/blob/29a1356669cb3c3ca2a6b22bced5047912072804/tests/python/test_shap.py#L64, so much easier, just sum all the contributions for each class for the scores in `score_weights`, but before confirm that I'm right and that the sum of the contributions is equal to the prediction for the document.
- Probably the rest could keep being the same (not sure). If that is the case, then what you would have to do is just replace this two lines https://github.com/AshwinB-hat/eli5/blob/master/eli5/xgboost.py#L195-L196 with some code based in `_xgd_feature_importances` that gives the desired `score_weights`. After that, all existing test should keep working. 

Sorry for saying you something wrong the other day. Hope now I have said something that makes more sense :-)

File: eli5/_shap_explanation_summary.py (hunk 1 above)
Comment (AshwinB-hat) [reply eli5/_shap_explanation_summary.py]: Yes! will do so. I was trying to mimic a module.

File: eli5/shap_xgboost.py (hunk 3 above)
Comment (AshwinB-hat) [reply eli5/shap_xgboost.py]: @ivanprado ,I have integrated xgboost. Over the next two days, I will try to finish `catboost` and `lightgbm`

File: eli5/xgboost.py [hunk 4]
```diff
@@ -413,3 +540,13 @@ def _missing_values_set_to_nan(values, missing_value, sparse_missing):
     if not np.isnan(missing_value):
         values[values == missing_value] = np.nan
     return values
+
+
+def shap_prediction_weights(booster, dmatrix, n_targets):
+    feature_weights = booster.predict(dmatrix, pred_contribs=True)
+    if n_targets > 1:
+        score_weights = [i for i in zip(np.sum(feature_weights, axis=2).squeeze(), feature_weights.squeeze())]
```
Comment (ivanprado) [inline eli5/xgboost.py]:
```suggestion
        score_weights = list(zip(np.sum(feature_weights, axis=2).squeeze(), feature_weights.squeeze()))
```

File: eli5/xgboost.py [hunk 5]
```diff
@@ -413,3 +540,13 @@ def _missing_values_set_to_nan(values, missing_value, sparse_missing):
...
+    else:
+        score_weights = [(np.sum(feature_weights), feature_weights.squeeze())]
+
+    return score_weights
```
Comment (ivanprado) [inline eli5/xgboost.py:554]: Could you add a test that checks that the score obtained from this function correspond with the prediction obtained from XGBoost? I mean, this is similar to assert that the sum of weights actually corresponds with the prediction that XGBoost returns.

File: eli5/xgboost.py [hunk 6]
```diff
@@ -413,3 +540,13 @@ def _missing_values_set_to_nan(values, missing_value, sparse_missing):
     if not np.isnan(missing_value):
         values[values == missing_value] = np.nan
     return values
+
+
+def shap_prediction_weights(booster, dmatrix, n_targets):
```
Comment (ivanprado) [inline eli5/xgboost.py]: What do you think about adding `_` to the name of this function to make it private?

**Comment by ivanprado** [comment]
Hi @AshwinB-hat . I left some minor comments, but implementation for explain_predictions in xgboost seems good :-). Congrats. I see that we preserve calling `get_scores` in xgboost for explain weights. I guess this is right, as this is the default implementation, but wonder if it would be better to replace by an implementation using SHAP (I mean, using the mean of SHAP values for all predictions). Maybe even `xgboost.get_scores` are doing so, no idea. So I have just this doubt, what do you think?

File: eli5/xgboost.py (hunk 5 above)
Comment (AshwinB-hat) [reply eli5/xgboost.py:554]: @ivanprado  Yes! i rechecked and that is exactly what happens. Do I add a test in `test_xgboost` or just an assert in `xgboost`?

**Comment by ivanprado** [comment]
Also, I didn't check the ligthgbm implementation as I'm waiting until you finish it.

File: eli5/xgboost.py (hunk 6 above)
Comment (AshwinB-hat) [reply eli5/xgboost.py]: @ivanprado Yes. It should be a private function. Adding with the next commit. Thanks

**Comment by AshwinB-hat** [comment]
> Hi @AshwinB-hat . I left some minor comments, but implementation for explain_predictions in xgboost seems good :-). Congrats. I see that we preserve calling `get_scores` in xgboost for explain weights. I guess this is right, as this is the default implementation, but wonder if it would be better to replace by an implementation using SHAP (I mean, using the mean of SHAP values for all predictions). Maybe even `xgboost.get_scores` are doing so, no idea. So I have just this doubt, what do you think?

I thought of this initially hence was averaging over all shap values. But `explain_weights` does not take the dataset as an argument. If I will have to average over all the predictions I will need to take the dataset as a parameter in `explain_weights` which it currently does not do!
What do you suggest? @ivanprado

File: eli5/xgboost.py (hunk 5 above)
Comment (ivanprado) [reply eli5/xgboost.py:554]: Having it in the test would be enough I guess.

**Comment by AshwinB-hat** [comment]
> Also, I didn't check the ligthgbm implementation as I'm waiting until you finish it.

Yes! I finished the Implementation but did not push the code since one existing test case was not passing. Will try to resolve and push by today with the suggested changes.

**Comment by AshwinB-hat** [comment]
@ivanprado  I have made the suggested changes and included LightGBM SHAP support which is passing all the existing test cases.

File: eli5/xgboost.py [hunk 7]
```diff
@@ -217,6 +214,136 @@ def explain_prediction_xgboost(
      )
 
 
+@explain_prediction.register(XGBClassifier)
+@explain_prediction.register(XGBRegressor)
+@explain_prediction.register(Booster)
+def explain_shap_prediction_xgboost(
```
Comment (ivanprado) [inline eli5/xgboost.py:220]: `explain_prediction_xgboost` and `explain_shap_prediction_xgboost` shares a lot of lines of code. Avoid code duplication for example by creating a function that accepts as parameter a function to use for retrieving the weights.

File: eli5/lightgbm.py [hunk 8]
```diff
@@ -151,6 +149,88 @@ def get_score_weights(_label_id):
      )
 
 
+@explain_prediction.register(lightgbm.LGBMClassifier)
+@explain_prediction.register(lightgbm.LGBMRegressor)
+def explain_shap_prediction_lightgbm(
```
Comment (ivanprado) [inline eli5/lightgbm.py:154]: Same here, `explain_shap_prediction_ligthgbm` and `explain_prediction_lightgbm` shares a lot of lines of code. Avoid the duplication of code as suggested for xgboost.

File: eli5/lightgbm.py [hunk 9]
```diff
@@ -278,3 +358,30 @@ def _target_feature_weights(feature_weights_dict, num_features, bias_idx):
 
 def _get_score(feature_weights_dict):
     return sum(feature_weights_dict.values())
+
+
+def _get_shap_prediction_feature_weights(lgb, X, n_targets):
```
Comment (ivanprado) [inline eli5/lightgbm.py:363]: I didn't dig into this function very much yet, but I have the intuition that using `np.reshape` function could simplify the code.

**Comment by ivanprado** [comment]
Good work @AshwinB-hat. Could you create a test that verifies that explanations make sense? I was thinking about creating a dataset with some features following some pattern (for example using periodic functions of different amplitude) and then using one of the features as target value. Train a model and then check that the output of explanation makes sense for a particular prediction, that is, that the most important feature is the target feature and that rest has very low importance. What do you think?

File: eli5/xgboost.py (hunk 7 above)
Comment (AshwinB-hat) [reply eli5/xgboost.py:220]: @ivanprado I did not make a function, since I thought that previous `explain_prediction` will be deprecated. Now if we intend to keep it, I have a suggestion. So by default we can have shap prediction but maybe adding an extra parameter to let the user choose between the type of explanation. What do you think? @ivanprado

File: eli5/lightgbm.py (hunk 9 above)
Comment (AshwinB-hat) [reply eli5/lightgbm.py:363]: @ivanprado Yes will make the required changes.

**Comment by AshwinB-hat** [comment]
> 
> 
> Good work @AshwinB-hat. Could you create a test that verifies that explanations make sense? I was thinking about creating a dataset with some features following some pattern (for example using periodic functions of different amplitude) and then using one of the features as target value. Train a model and then check that the output of explanation makes sense for a particular prediction, that is, that the most important feature is the target feature and that rest has very low importance. What do you think?

Its a nice idea to be able to evaluate the predictions. This will come in handy in PhaseTwo where I have to implement the custom SHAP mechanism. Although my doubt is which periodic function do I use which has a significant number of features? @ivanprado

File: eli5/xgboost.py (hunk 7 above)
Comment (ivanprado) [reply eli5/xgboost.py:220]: Sounds good to me.

Comment (ivanprado) [reply eli5/xgboost.py:220]: Besides test should be then updated to run them for both versions just by adding a new `pytest.mark.parametrize`

File: eli5/lightgbm.py (hunk 9 above)
Comment (ivanprado) [reply eli5/lightgbm.py:363]: Any progress in that?

**Comment by ivanprado** [comment]
Hi @AshwinB-hat, I'm back again :-). I did a quick review of the PR. I run the Titanic notebook with your code for XGBoost and results looks great!. 

I wonder if you could extend it/create a similar one to show the same results but using ligthgbm. It would be a way to check the results of your changes and at the same time will be useful as documentation.

Remaining tasks for xgboost/ligthgbm:

- [ ] New parameter to switch to old explanations
- [ ] Test for both types of explanations (old/shap)
- [ ] Update documentation
- [ ] Code simplification if possible by using more numpy primitives
- [ ] Notebook for ligthgbm

File: eli5/xgboost.py (hunk 7 above)
Comment (AshwinB-hat) [reply eli5/xgboost.py:220]: @ivanprado  Yes then I will include a parameter and include parametrize

File: eli5/lightgbm.py (hunk 9 above)
Comment (AshwinB-hat) [reply eli5/lightgbm.py:363]: Yes, I changed that in the new commit. I will recheck
